pkg/components/rook/component_conversion_test.go

```diff
@@ -0,0 +1,232 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rook_test
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kinvolk/lokomotive/pkg/components"
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+)
+
+func renderManifests(t *testing.T, configHCL string) map[string]string {
+	name := "rook"
+
+	component, err := components.Get(name)
+	if err != nil {
+		t.Fatalf("Getting component %q: %v", name, err)
+	}
+
+	body, diagnostics := util.GetComponentBody(configHCL, name)
+	if diagnostics != nil {
+		t.Fatalf("Getting component body: %v", diagnostics)
+	}
+
+	diagnostics = component.LoadConfig(body, &hcl.EvalContext{})
+	if diagnostics.HasErrors() {
+		t.Fatalf("Valid config should not return an error, got: %s", diagnostics)
+	}
+
+	ret, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	return ret
+}
+
+func getOperatorDeployment(t *testing.T, m map[string]string) *appsv1.Deployment {
+	deployStr, ok := m["rook-ceph/templates/deployment.yaml"]
+	if !ok {
+		t.Fatalf("Operator deployment config not found")
+	}
+
+	deploy := &appsv1.Deployment{}
+	if err := yaml.Unmarshal([]byte(deployStr), deploy); err != nil {
+		t.Fatalf("Unmarshaling manifest: %v", err)
+	}
+
+	return deploy
+}
+
+// nolint:funlen
+func TestConversion(t *testing.T) {
+	configHCL := `
+component "rook" {
+  enable_monitoring = true
+
+  node_selector = {
+    "storage_node_key1" = "storage_node_value1"
+    "storage_node_key2" = "storage_node_value2"
+  }
+
+  toleration {
+    key      = "toleration_key1"
+    operator = "Equal"
+    value    = "toleration_value1"
+    effect   = "NoSchedule"
+  }
+
+  toleration {
+    key      = "toleration_key2"
+    operator = "Equal"
+    value    = "toleration_value2"
+    effect   = "NoSchedule"
+  }
+
+  agent_toleration_key    = "agent_toleration_key"
+  agent_toleration_effect = "NoSchedule"
+
+  discover_toleration_key    = "discover_toleration_key"
+  discover_toleration_effect = "NoSchedule"
+
+  csi_plugin_node_selector = {
+    "node_key1" = "node_value1"
+    "node_key2" = "node_value2"
+  }
+
+  csi_plugin_toleration {
+    key = "other_node_key1"
+    operator = "Equal"
+    value = "other_node_value1"
+    effect = "NoSchedule"
+  }
+}
+`
+
+	m := renderManifests(t, configHCL)
+	if len(m) == 0 {
```

Comment:
With such code, this condition should be put in separate test IMO.

Reply:
If I put it in a separate function, how do I stop when such an error is encountered? If this fails as in `len(m)` is zero then this test will fail on error: `Operator deployment config not found`.

So catching this before we proceed further makes sense and gives up precise error.

Reply:
Hmm, yeah, I this sort of make sense. But other tests will fail gracefully if we remove this condition and we assume all rendering fails? I wonder how this should be handled in Go :thinking: I guess it's fine then :smile:

Reply:
The way the tests are currently structured, rendering the manifest is not the actual **check** in the unit test but rather a preliminary step. What @invidian is talking about sounds like treating a zero-length manifest as a **test case** on its own (which is valid, but possibly a different "category" than the tests in this function).